docs/adr/reconstruction.md

```diff
@@ -0,0 +1,286 @@
+
+## Specification
+
+### Performance Requirements
+
+1. Base Scenario Support:
+    - 32MB block size
+    - Minimum required light nodes for 32MB blocks
+    - Network of 50+ full nodes
+
+2. Performance Metrics (in order of priority):
+    - System stability
+    - Reconstruction throughput (blocks/second)
+    - Per-block reconstruction time
+
+3. Scalability Dimensions:
+    - Block size scaling
+    - Node count scaling
+
+
+
+### Protocol Flow
+
+```
+1. Connection Establishment
+   Client                              Server
+      |---- Handshake (node type) ----->|
+      |<---- Handshake (node type) -----|
+
+2. Bitmap Subscription
+   Client                                   Server
+      |---- Subscribe to bitmap ------------->|
+      |<---- Initial bitmap ------------------|
+      |<---- Updates  ------------------------|
+      |<---- End updates(full eds/max samples-|
+
+3. Data Request
+   Client                              Server
+      |---- Request(bitmap) ----------->|
+      |<---- [Samples + Proof] parts ---|
+```
+
+### Bitmap Implementation
+
+The protocol utilizes Roaring Bitmaps for efficient bitmap operations and storage. Roaring Bitmaps provide several advantages for the reconstruction protocol:
+
+1. Efficient Operations
+    - Fast logical operations (AND, OR, XOR)
+    - Optimized for sparse and dense data
+    - Memory-efficient storage
+
+2. Implementation Benefits
+    - Native support for common bitmap operations
+    - Optimized serialization
+    - Efficient iteration over set bits
+    - Support for rank/select operations
+
+3. Performance Characteristics
+    - O(1) for most common operations
+    - Compressed storage format
+    - Efficient memory utilization
+    - Fast bitmap comparisons
+
+
+### Core Protocol Components
+
+#### 1. Connection Management
+
+Handshakes, Node Type Identification
+- Node type must be declared during connection handshake by each of connection nodes
+- Active connection state maintenance on full nodes. List should be maintained by node type
+- Connection state should allow subscription for updated
+
+#### 2. Sample Bitmap Protocol
+
+2.1. Bitmap Subscription Flow
+
+- Requested by block height
+- Implements one-way stream for bitmap updates
+- Full nodes: Stream until complete EDS bitmap
+- Light nodes: Stream until max sampling amount reached
+- Include end-of-subscription flag for easier debugging
+
+2.2. Update Requirements
+- Minimum update frequency: every 10 seconds
+- Server-side update triggers:
+    - Time-based: Every 5 seconds. Default in base implementation.
+    - Optional: Change-based with threshold
+- Penalty system for delayed updates
+
+#### 3. Reconstruction State Management
+
+3.1. Global State Components
+```go
+type RemoteState struct {
+    coords [][]peers    // Peer lists by coordinates
+    rows []peers        // Peers with row data
+    cols []peers        // Peers with column data
```

Comment:
how is this actually going to be structured? 

you keep a []peer.ID per which row they claim to have? So index 0 []peer.ID will be peers that claim to have row 0?

Reply:
Yes. The idea is to have efficient look up for peers who have full row or full col.